http-client/src/main/java/io/micronaut/http/client/MatchingFilterResolver.java

```diff
@@ -0,0 +1,171 @@
+        io.micronaut.http.HttpMethod method = request.getMethod();
+        for (HttpClientFilter filter : filters) {
+            if (filter instanceof Toggleable && !((Toggleable) filter).isEnabled()) {
+                continue;
+            }
+
+            Class<? extends Annotation>[] clientDeclaredFilterAnnotations = (Class<? extends Annotation>[]) annotationMetadata
+                    .getAnnotationTypesByStereotype(FilterAnnotation.class)
+                    .stream()
+                    .toArray(Class[]::new);
+
+            Class<? extends Annotation>[] declaredFilterAnnotations = (Class<? extends Annotation>[]) annotationMetadataResolver.resolveMetadata(filter)
+                    .getAnnotationTypesByStereotype(FilterAnnotation.class)
+                    .stream()
+                    .toArray(Class[]::new);
+
+            Optional<AnnotationValue<Filter>> filterAnnotationMetadataOpt = annotationMetadataResolver.resolveMetadata(filter)
+                    .findAnnotation(Filter.class);
+
+            FilterProperties annotationFilterProperties = filterAnnotationMetadataOpt
+                    .map(f -> new FilterProperties(
+                            f.stringValues("patterns"),
+                            f.get("methods", io.micronaut.http.HttpMethod[].class, null),
+                            f.stringValues("serviceId"),
+                            declaredFilterAnnotations)
+                    )
+                    .orElse(FilterProperties.EMPTY_FILTER_PROPERTIES);
+
+            FilterProperties instanceProperties = Optional.of(filter)
+                    .filter(f -> f instanceof MatchingFilterResolver)
+                    .map(f -> (MatchingHttpClientFilter) f)
+                    .map(MatchingHttpClientFilter::getFilterProperties)
+                    .orElse(FilterProperties.EMPTY_FILTER_PROPERTIES);
+
+            for (Class<? extends Annotation> stereotype : instanceProperties.getStereotypes()) {
+                if (!annotationMetadataResolver.resolveMetadata(stereotype).hasStereotype(FilterAnnotation.class)) {
+                    LOG.warn("Filter stereotype annotation {} is not marked with @FilterAnnotation that makes it completely ignored.", stereotype.getSimpleName());
+                }
+            }
+
+            FilterProperties filterProperties = instanceProperties.merge(annotationFilterProperties);
+
+            BooleanSupplier pathMatches = () -> filterByPath(filterProperties.getPatterns(), requestPath);
+            BooleanSupplier methodMatches = () -> filterByMethod(filterProperties.getMethods(), method);
+            BooleanSupplier serviceIdMatches = () -> filterByServiceId(filterProperties.getServiceId());
+            BooleanSupplier filterStereotypeMatched = () -> filterByStereotype(filterProperties.getStereotypes(), clientDeclaredFilterAnnotations);
+
+            boolean filterMatched = Stream.of(pathMatches, methodMatches, serviceIdMatches, filterStereotypeMatched)
```

Comment:
Whilst this looks pretty the performance will be worse than just doing a `for` loop and breaking out at the first match since you have to construct 4 lambdas and them a stream from them. Lambdas are nice but sometimes not as efficient.

Reply:
Replaced stream with more fermormnat if statement
```
if (isPathMatches(filterProperties.getPatterns(), requestPath)
                    && isMethodMatches(filterProperties.getMethods(), method)
                    && isServiceIdMatches(filterProperties.getServiceId())
                    && isStereotypeMatches(filterProperties.getStereotypes(), clientDeclaredFilterAnnotations)
            ) {
                filterList.add(filter);
            }
```